### PR TITLE
[azcosmos] unblock cross-region failover from synchronous gem.Update

### PR DESCRIPTION
**Pull request by tvaron3**
## What & Why

Three retry paths in the Cosmos client retry policy (`cosmos_client_retry_policy.go`) used to gate the cross-region retry on a successful synchronous `gem.Update(ctx, true)` against the account's global endpoint. During a regional outage the global FQDN typically resolves to the same regional FE pool we just marked unavailable (Cosmos accounts in a region share FEs), so the forced refresh stalls on a connect timeout and surfaces that error instead of attempting the failover.

This PR also fixes two interlocking bugs in `attemptRetryOnNetworkError` that prevented the failover from actually routing to another region even when the refresh succeeded, and tightens a few related correctness issues.

### Changes

- **`attemptRetryOnNetworkError` (connection error)** — dropped the forced `gem.Update`. `MarkEndpointUnavailable*` already calls `locationCache.updateLocked` which recomputes `readEndpoints` / `writeEndpoints` with the unavailable endpoint demoted; `invalidate()` arms the next non-force `Update` for real topology changes. Also left `retryCount` at 0 after the failover — `getPrefAvailableEndpointsLocked` demotes unavailable endpoints to the *tail* of the route list rather than removing them, so `retryCount=1` (after `retryCount += 1`) re-indexed back to the just-marked unhealthy endpoint via `ResolveServiceEndpoint(1 % 2)`.
- **`attemptRetryOnEndpointFailure` (HTTP 403 / `WriteForbidden`)** — replaced the blocking `gem.Update` with a fire-and-forget `asyncForceRefreshGEM` helper. The helper uses a `CompareAndSwap` gate so a retry storm cannot leak a goroutine per attempt, runs on `context.Background()` so a near-expired caller deadline cannot abort the refresh, and has panic recovery + error logging.
- **`attemptRetryOnRequestTimeout` (HTTP 408)** — removed both `MarkEndpointUnavailableForRead` and the forced `gem.Update`. A 408 is a per-request signal (the gateway accepted the request but did not finish in time), not regional unhealthiness; demoting the whole region for `unavailableLocationExpirationTime` based on one slow request would penalize concurrent reads. The outer loop's `retryCount++` already advances the location index for the cross-region retry.
- **URL-key normalization in `locationCache`** — `MarkEndpointUnavailable*` was called with the full request URL (path/query/RawPath), but the unavailability map was keyed by `url.URL` struct equality while `availReadEndpointsByLocation` / `availWriteEndpointsByLocation` only stored scheme+host. Marks were silently written under a key nothing else looked up, so `isEndpointUnavailableLocked` always returned false and the demote was a no-op. Now normalized to scheme+host via a new internal `endpointKey` helper.
- **`locationCache.resolveServiceEndpoint`** — now takes `mapMutex.RLock`. Reads of `locationInfo` / `enableMultipleWriteLocations` race the writes in `update`/`updateLocked`, which the new async refresh paths surface deterministically under `go test -race`.

### Repro that motivated the fix

Reproduced live by running `azcosmos_perf` in AKS with `iptables OUTPUT DROP` rules applied via `kubectl debug --profile=netadmin` to block the workload account's `westus2` FE IPs (`20.42.169.199`, `20.42.170.147`, `20.9.156.133`).

Account topology:
- Workload account: write=`westus2`, reads=`[westus2, westus]`
- Client args: `--application-region westus2 --preferred-regions westus`

For 10 minutes:
- `ss -tan` inside the pod showed every goroutine in `SYN_SENT` to a blocked `westus2` IP. **Zero TCP attempts to `40.112.241.99` (the `westus` replica)**.
- Per-op error rate ≈ 15/min/op across `concurrency=50` → ~18–22 s wall-clock per failed op, matching the budgeted same-region retries (3 × (5s + 1s backoff)) plus one failover attempt whose `gem.Update` itself timed out.
- Effective throughput went to zero and stayed there for the full window.

After the fix, the cache mutation from `MarkEndpointUnavailable*` is enough for the next iteration's `ResolveServiceEndpoint` to return `westus`, and the cross-region retry actually executes.

## Regression tests

- `TestConnectionErrorReadFailsOverWhenGlobalEndpointIsUnreachable` — pins the GEM mock to `net.DNSError` so any forced refresh would fail; asserts the 5th-attempt 200 from the failover server is returned. Pre-fix the request surfaces a DNS error instead.
- `TestFix3a_WriteRetrySucceedsWhenGEMRefreshFails` — 403 retry returns `(true, nil)` even when GEM transport returns 500 on every call.
- `TestFix3b_RequestTimeoutDoesNotMarkEndpointUnavailable` — after a 408 retry, `locationUnavailabilityInfoMap` is empty and the GEM transport was never invoked.
- `TestFix3c_RequestTimeoutDoesNotBlockOnStalledGEM` — with a 10-minute hanging GEM transport, `attemptRetryOnRequestTimeout` returns within 5 s.
- `TestFix3d_AsyncRefreshCASGateCoalescesRetryStorm` — 50 concurrent `asyncForceRefreshGEM()` calls produce far fewer than `concurrency/2` HTTP refreshes.
- `TestFix3e_AsyncRefreshIgnoresCallerContextCancellation` — an already-cancelled caller context still triggers a successful background refresh (proves `context.Background()`).
- `TestFix3_WriteRetryForceRefreshesGEM` updated to use `require.Eventually` since the refresh is now async and the singleflight + CAS gate legitimately coalesce concurrent forced refreshes.

## Backwards compatibility

- No public API changes.
- The single-master-write branch of `attemptRetryOnNetworkError` already had a comment explicitly noting that no `gem.Update` is needed here (added in #26815). This change extends the same reasoning to the read / multi-master-write failover branch and to the 403 / 408 paths.

## CHANGELOG

Entry added under `1.5.0-beta.7` → `Bugs Fixed`.

## Testing

```
go test -count=1 -timeout 300s -race ./...
ok  github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos    71.863s
```

### CODE REVIEW

File: sdk/data/azcosmos/cosmos_client_retry_policy_test.go [hunk 1]
```diff
@@ -601,6 +601,75 @@ func TestConnectionErrorReadFailsOverAfterThreeSameRegionAttempts(t *testing.T)
...
+// to the same regional FE pool that has been blocked (the common case for
+// single-region writes — global FQDN points at the write region's FE).
+//
+// Before the fix, attemptRetryOnNetworkError forced a synchronous
+// gem.Update(ctx, true) after MarkEndpointUnavailable*; the topology
+// refresh dialed the unreachable global endpoint, hit a connect timeout,
+// and the failover bailed with that error — surfacing the original
+// connection failure to the caller and never trying the next preferred
+// region whose state was already populated in the cache by
+// MarkEndpointUnavailable*.
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/data/azcosmos/cosmos_client_retry_policy_test.go]: Addressed in 4c19925ef8. The test injects `net.DNSError` as a deterministic, no-sleep stand-in for the connect-timeout failure mode that happens in production when the global FQDN resolves to a blocked regional FE pool — both cause `gem.Update(ctx, true)` to return `(false, err)`, which is the only behavior the test relies on. Updated the doc block and the inline comment on `gemServer.SetError` to spell that out so a reader of the test in isolation sees why DNSError stands in for the production failure.

File: sdk/data/azcosmos/cosmos_client_retry_policy_test.go [hunk 2]
```diff
@@ -601,6 +601,75 @@ func TestConnectionErrorReadFailsOverAfterThreeSameRegionAttempts(t *testing.T)
...
+//
+// The fix removes the forced refresh; the cache mutation from
+// MarkEndpointUnavailable* is enough for the next ResolveServiceEndpoint
+// to pick the failover region.
+func TestConnectionErrorReadFailsOverWhenGlobalEndpointIsUnreachable(t *testing.T) {
+	srv, srvClose := mock.NewTLSServer()
+	defer srvClose()
+
+	defaultEndpoint, err := url.Parse(srv.URL())
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (tvaron3) [reply sdk/data/azcosmos/cosmos_client_retry_policy_test.go]: Considered, but not adopted. The other call sites of `setupRetryPolicyTestClientOpts` share a single mock server + `CreateMockLC` (which points every region at the same `defaultEndpoint`); this test deliberately needs (a) two distinct backend mock servers wired through a host-routing transport so the failover's destination is actually observable, (b) a separate GEM mock server with its own injected error, and (c) a real location cache with two distinct regional endpoints. Folding all of that into the helper as configurable knobs would obscure it for the existing simple callers without removing meaningful duplication — the setup here is genuinely a different shape. Happy to revisit if a second "two-region routing" test gets added later.

**Review by tvaron3 — COMMENTED** [review]
**PR Deep Review summary**

I found three blocking correctness risks and several targeted test/robustness gaps. The main concern is that the new demote-to-tail routing only works when the next retry indexes `0`; the 403 path still increments `retryCount`, and the connection-error path does not reset a pre-existing non-zero `retryCount`. I also found remaining unlocked location-cache readers exposed by the new async refresh path.

I could not run local Go tests because `go` is not installed in this environment (`go: command not found`).

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 3]
```diff
@@ -331,10 +406,14 @@ func (p *clientRetryPolicy) attemptRetryOnEndpointFailure(req *policy.Request, i
...
-		return false, err
-	}
+	// Fire-and-forget the topology refresh. We do NOT block the retry on
+	// its outcome: MarkEndpointUnavailable* has already invalidated the
+	// GEM cache and demoted the bad endpoint locally, so the next
+	// ResolveServiceEndpoint will pick the failover region whether or
+	// not the metadata refresh succeeds. Blocking here would surface a
+	// transient metadata failure to the caller and skip the very
+	// cross-region retry this function is supposed to perform.
+	p.asyncForceRefreshGEM()
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go]: **🔴 Blocking** · Correctness: Retry Routing

**403 retries still index back into the demoted endpoint**

This path marks the current endpoint unavailable, which recomputes the route list with that endpoint moved to the tail. But after `attemptRetryOnEndpointFailure` returns, the outer loop unconditionally does `retryContext.retryCount += 1`. For a two-region account, the list becomes `[healthy, unhealthy]`, then `retryCount == 1` resolves right back to the unhealthy tail entry. That is the same routing bug this PR fixes for connection errors, but it still exists for 403/`WriteForbidden`.

Why it matters: during a write-region failover, the retry can keep hitting the endpoint that just returned `WriteForbidden`, while the async GEM refresh is only opportunistic and may fail or lag. Consider resetting the index used after a demote (for example set `retryCount` so the outer increment leaves it at `0`, or have the outer loop skip the generic increment for mark-and-demote retries).

```go
// After MarkEndpointUnavailable* demotes the current endpoint, the next
// ResolveServiceEndpoint call must use index 0, not the generic +1.
retryContext.retryCount = -1 // outer loop increments to 0
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 4]
```diff
@@ -299,15 +354,35 @@ func (p *clientRetryPolicy) attemptRetryOnNetworkError(req *policy.Request, kind
...
+	// (false, err) and surface a connection error to the caller without
+	// ever attempting the cross-region retry.
+	//
+	// A connection error indicates regional unhealthiness, not a topology
+	// change, so skipping the synchronous metadata refresh here is safe.
+	// The invalidate() inside MarkEndpointUnavailable* ensures the next
+	// background or non-force Update will pick up any actual topology
+	// changes when the global endpoint is reachable again.
 
 	retryContext.sameRegionRetryCount = 0
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go:395]: **🔴 Blocking** · Correctness: Retry State

**Connection-error failover must reset an existing non-zero retryCount**

The new logic intentionally avoids incrementing `retryCount`, but it does not reset it. That only works for tests where the connection-error chain starts at `retryCount == 0`. A real request can reach this path after an earlier HTTP-status retry has already incremented `retryCount` (for example 408/503 followed by a transport failure on the second region). After `MarkEndpointUnavailable*` demotes the failed endpoint to the tail, leaving `retryCount == 1` makes the next resolve pick that tail entry again.

Why it matters: mixed failure sequences can still fail over to the region that was just marked unhealthy. Reset the index explicitly when relying on the demote-to-tail invariant.

```go
retryContext.sameRegionRetryCount = 0
retryContext.retryCount = 0
retryContext.crossRegionFailoverDone = true
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 5]
```diff
@@ -145,6 +145,14 @@ func (lc *locationCache) updateLocked(writeLocations []accountRegion, readLocati
 }
 
 func (lc *locationCache) resolveServiceEndpoint(locationIndex int, resourceType resourceType, isWriteOperation, useWriteEndpoint bool) url.URL {
+	// Take a read lock for the duration of endpoint resolution. The
+	// fields read here (locationInfo, enableMultipleWriteLocations) are
+	// rewritten atomically under mapMutex.Lock() by update/updateLocked,
+	// and a concurrent forced refresh (e.g. from the retry policy's
+	// asyncForceRefreshGEM or the GEM policy's background refresh) can
+	// race with us without this lock.
+	lc.mapMutex.RLock()
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_location_cache.go:154]: **🔴 Blocking** · Correctness: Data Race

**The new lock fixes resolveServiceEndpoint, but async refresh still races other locationInfo readers**

The `RLock` here protects `resolveServiceEndpoint`, but the new async refresh path can still call `locationCache.update()` while other data-plane code reads the same fields without a lock. Two examples in the same request flow are `GetEndpointLocation` → `locationCache.getLocation` (iterates `availWriteEndpointsByLocation` / `availReadEndpointsByLocation`) and `attemptRetryOnSessionUnavailable` (reads `locationInfo.availReadLocations` / `availWriteLocations`).

Why it matters: this PR makes refreshes happen concurrently with the retry loop more often, so leaving those readers unlocked can still fail `go test -race` or panic on concurrent map access. Please extend the same locking/snapshot pattern to every `locationInfo`/`enableMultipleWriteLocations` reader, not just endpoint resolution.

```go
func (lc *locationCache) getLocation(endpoint url.URL) string {
    lc.mapMutex.RLock()
    defer lc.mapMutex.RUnlock()
    return lc.getLocationLocked(endpoint)
}
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 6]
```diff
@@ -11,16 +11,71 @@ import (
...
+		// here so a panic in the GEM pipeline does not bring down the
+		// host via this detached goroutine.
+		defer func() {
+			if r := recover(); r != nil {
+				log.Writef(azlog.EventResponse,
+					"panic in azcosmos retry-policy async GEM refresh: %v\n%s",
+					r, debug.Stack())
+			}
+		}()
+		if err := p.gem.Update(context.Background(), true); err != nil {
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go]: **🟡 Recommendation** · Reliability: Refresh Coalescing

**A forced async refresh can be swallowed by an older in-flight refresh**

`gem.Update(..., true)` still coalesces with any existing `gem.inflight`, even if that flight started before `MarkEndpointUnavailable*` invalidated the GEM. In that case this goroutine waits for the stale flight and exits without starting a post-invalidation refresh. `invalidationGen` prevents the stale flight from advancing timestamps, but this force-refresh caller does not loop and become the leader for the fresh attempt.

Why it matters: after a 403/`WriteForbidden`, the intended topology refresh may be delayed until some later request notices `ShouldRefresh()`, which weakens the “kick off a forced refresh now” guarantee. Consider teaching `Update(forceRefresh=true)` to start a new flight after waiting on a pre-invalidation one, or have `asyncForceRefreshGEM` retry once when the waited flight did not satisfy the invalidation.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go (hunk 3 above)
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go]: **🟡 Recommendation** · Performance: Metadata Storm

**Sequential 403s can force-refresh GEM repeatedly after the endpoint is already marked unavailable**

`MarkEndpointUnavailable*` already knows whether this is a new unavailability event and only invalidates GEM on the first mark. This call ignores that signal and always schedules a forced refresh. Because `forceRefresh=true` bypasses `refreshTimeInterval`, a sustained stream of non-overlapping 403s can still issue one `GetDatabaseAccount` per retry/backoff interval even though the endpoint is already in the unavailability window.

Why it matters: PR #26815 fixed `GetDatabaseAccount` storms; this adds a new sequential storm shape (the CAS gate only coalesces overlapping calls). Consider returning `wasAlreadyUnavailable` from the GEM wrapper, or otherwise throttling forced async refreshes so only the first new mark forces topology refresh while repeated marks rely on the existing invalidation/throttle.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 7]
```diff
@@ -331,10 +406,14 @@ func (p *clientRetryPolicy) attemptRetryOnEndpointFailure(req *policy.Request, i
...
+	// Fire-and-forget the topology refresh. We do NOT block the retry on
+	// its outcome: MarkEndpointUnavailable* has already invalidated the
+	// GEM cache and demoted the bad endpoint locally, so the next
+	// ResolveServiceEndpoint will pick the failover region whether or
+	// not the metadata refresh succeeds. Blocking here would surface a
+	// transient metadata failure to the caller and skip the very
+	// cross-region retry this function is supposed to perform.
+	p.asyncForceRefreshGEM()
 
 	time.Sleep(defaultBackoff * time.Second)
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go]: **🟡 Recommendation** · Reliability: Cancellation Handling

**403 backoff still ignores caller cancellation**

The connection-error and 408 paths now use `sleepWithContext`, but this 403 path still uses `time.Sleep`. If the caller deadline expires during the one-second backoff, the retry policy waits the full second before returning or attempting the next request.

Why it matters: one of the retry-policy guarantees added in the recent work is not consuming the caller's end-to-end timeout budget with internal backoffs. This path should preserve that behavior too.

```suggestion
	if sleepErr := sleepWithContext(req.Raw().Context(), defaultBackoff*time.Second); sleepErr != nil {
		return false, sleepErr
	}
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 8]
```diff
@@ -299,15 +354,35 @@ func (p *clientRetryPolicy) attemptRetryOnNetworkError(req *policy.Request, kind
...
-	// Force a refresh so the new unavailability is reflected in
-	// readEndpoints / writeEndpoints for both this request and any
-	// concurrent requests racing through resolveServiceEndpoint.
-	if err := p.gem.Update(req.Raw().Context(), true); err != nil {
-		return false, err
-	}
+	// Do NOT force a topology refresh here. MarkEndpointUnavailable* has
+	// already recomputed locationInfo.readEndpoints / writeEndpoints with
+	// the bad endpoint demoted, so the next ResolveServiceEndpoint call
+	// (with retryCount+1) will pick the next preferred region on its own.
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go]: **🟡 Recommendation** · Maintainability: Misleading Comment

**The comment still says the next resolve uses retryCount+1**

This rationale now contradicts the implementation: the function intentionally does not increment `retryCount`, and the error path does not have the outer HTTP-status `retryCount++`. The next `ResolveServiceEndpoint` uses the same index, which is why the code depends on that index being/resetting to `0`.

Why it matters: this is the core invariant of the fix. A future maintainer following the comment could reintroduce the off-by-one behavior.

```suggestion
	// the bad endpoint demoted, so the next ResolveServiceEndpoint call
	// (using index 0 after this failover resets retryCount) will pick the
	// next preferred region on its own.
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_dbaccount_refresh_test.go [hunk 9]
```diff
@@ -186,12 +187,256 @@ func TestFix3_WriteRetryForceRefreshesGEM(t *testing.T) {
...
+		lastUpdateTime:      time.Now(),
+	}
+	retry := &clientRetryPolicy{gem: gem}
+
+	req, err := azruntime.NewRequest(context.Background(), http.MethodPost, defaultEndpoint.String())
+	require.NoError(t, err)
+
+	rc := &retryContext{}
+	shouldRetry, err := retry.attemptRetryOnEndpointFailure(req, true, rc)
+	require.NoError(t, err, "GEM refresh failures must not surface to the retry caller")
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_dbaccount_refresh_test.go:261]: **🟡 Recommendation** · Tests: False Negative

**This failure-path test passes even if the async refresh never runs**

`attemptRetryOnEndpointFailure` now returns before the goroutine reaches `gem.Update`, and this test only asserts the immediate `(true, nil)` result. If `asyncForceRefreshGEM` became a no-op, or returned early because the CAS gate was stuck, this test would still pass.

Why it matters: the regression is specifically that GEM failures must not block the retry, but we still need to prove the failing refresh path was actually exercised. Add an `Eventually` on `transport.count` after the call, mirroring the successful-refresh tests.

```go
require.Eventually(t, func() bool {
    return transport.count.Load() >= 1
}, 5*time.Second, 10*time.Millisecond,
    "the async GEM refresh should run and fail without surfacing to the retry caller")
```

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_dbaccount_refresh_test.go [hunk 10]
```diff
@@ -186,12 +187,256 @@ func TestFix3_WriteRetryForceRefreshesGEM(t *testing.T) {
...
-		"write retries on 403/WriteForbidden must force-refresh the GEM on every attempt")
+	// The refresh is fire-and-forget, so wait for the detached goroutines
+	// to land at least one HTTP call. We require >=1 (not ==writeRetries)
+	// because gem.Update's single-in-flight pattern plus the policy's own
+	// asyncRefreshPending gate legitimately coalesces concurrent forced
+	// refreshes into a single GetDatabaseAccount call.
+	require.Eventually(t, func() bool {
+		return transport.count.Load() >= 1
+	}, 5*time.Second, 10*time.Millisecond,
+		"write retries on 403/WriteForbidden must kick off a GEM refresh")
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_dbaccount_refresh_test.go]: **🟡 Recommendation** · Tests: Coverage Gap

**The “per attempt” refresh test now only requires one refresh total**

The test name/comment still says 403 retries kick off a GEM refresh per attempt, and the loop performs five sequential attempts. The assertion only requires `transport.count >= 1`, so a regression that refreshes once and then never refreshes again would pass. The CAS gate should coalesce overlapping calls, not permanently suppress later sequential refreshes after the first goroutine completes.

Why it matters: if the gate gets stuck or the code is changed to only refresh on the first 403, topology updates after later failover signals can be missed. Either tighten the assertion for sequential attempts (for example `>= 2` or `== writeRetries` if that is the contract), or rename the test to cover only “at least one async refresh is scheduled.”

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: sdk/data/azcosmos/cosmos_client_retry_policy_test.go [hunk 11]
```diff
@@ -598,7 +599,147 @@ func TestConnectionErrorReadFailsOverAfterThreeSameRegionAttempts(t *testing.T)
...
+//     endpoint at the tail — the failover attempt would hit the same
+//     dead region again.
+//  3. MarkEndpointUnavailable* was called with the full request URL
+//     (path, query, etc. included) but the unavailability map and the
+//     cache's per-region endpoint lookup were keyed by base URLs
+//     (scheme+host). The marks were therefore written under keys nothing
+//     else looked up, so isEndpointUnavailableLocked always returned false
+//     and the demote silently did nothing.
+//
+// The fix drops the forced refresh, leaves retryCount at 0 so the next
```
Comment (tvaron3) [inline sdk/data/azcosmos/cosmos_client_retry_policy_test.go]: **🟡 Recommendation** · Tests: Missing Edge Case

**The connection-error regression only covers retryCount starting at zero**

This test is valuable, but it proves the demote-to-tail fix only for a fresh request with `retryCount == 0`. The implementation does not reset `retryCount`; it only avoids incrementing it. If a request first consumes an HTTP-status retry and then hits the connection-error failover path, the stale non-zero index can still point at the demoted endpoint.

Why it matters: mixed failure sequences are plausible during outages (for example a 408/503 followed by a dial failure). Add a regression that seeds `retryCount` non-zero before connection-error failover, or drives the full pipeline through an HTTP retry followed by transport errors, and assert the failover host is the healthy one.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

**Review by kushagraThapar — APPROVED** [review]
Thanks for the rapid turnaround on top of #26815 and the excellent live-AKS repro narrative — `iptables OUTPUT DROP` + `ss -tan` showing zero TCP attempts to the replica IP is exactly the level of forensic detail that made the three interlocking bugs visible. The URL-key normalization catch was particularly nice. Core fix (`MarkEndpointUnavailable*` cache mutation is sufficient; synchronous refresh is the trap) reads cleanly and aligns with .NET on the demote-to-tail routing model. LGTM 🚀

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @tvaron3

**Review by kushagraThapar — APPROVED** [review]
